proxy_test.go

```diff
@@ -253,11 +258,16 @@ func TestReverseProxy_ServeHTTP1(t *testing.T) {
 			f: func(p *reverseProxy) *http.Response {
 				p.users["default"].maxConcurrentQueries = 1
 				p.users["default"].queueCh = make(chan struct{}, 1)
-				go makeHeavyRequest(p, time.Millisecond*100)
-				time.Sleep(time.Millisecond * 5)
-				go makeHeavyRequest(p, time.Millisecond*100)
-				time.Sleep(time.Millisecond * 5)
-				return makeHeavyRequest(p, time.Millisecond*20)
+				// we don't wait the requests to be handled by the fakeServer because one of them will be enqueued and not handled
```

Comment:
can you elaborate a bit about this particular case? enqueued where?

Reply:
This part is tricky, it took me some time to get it!

if you look at the logic in runHeavyRequestInGoroutine. It waits that the nbHeavyRequestsInflight reaches the number of requests asked. The nbHeavyRequestsInflight is incremented in the HandlerFunc of the fakeServer. This server simulates a clickhouse. In the tests, we have no way to know whether the goroutines have reach the part of the code linked to chproxy. This is why I put the increment on the fake clickhouse sever because if the query is currently sleeping into the fake clickhouse it means that it's currently handled by CHproxy. For some tests we need to be sure that the request in the goroutines is really handled by CHProxy, otherwise it can lead to race conditions .


But, on this specific test I trigger 2 goroutines with a queue and a max concurrent query at 1.
The first request reaches CHProxy https://github.com/ContentSquare/chproxy/blob/6dce06adb0a10ac260c84d07315fa6ae4cb18931/proxy_test.go#L732
Then CHProxy checks if the queries needs to be put in a queue, which is not the case for the first query
https://github.com/ContentSquare/chproxy/blob/6dce06adb0a10ac260c84d07315fa6ae4cb18931/proxy.go#L68
Then the query is sent to clickhouse which means in the test nbHeavyRequestsInflight is incremented

But for the 2nd request, it reaches CHProxy then is stopped by the queuing logic until the first request ends. Which means that nbHeavyRequestsInflight is first decremented by the first query then incremented by the second query. So the nbHeavyRequestsInflight will never reach 2 and therefore we can't rely on the logic inside runHeavyRequestInGoroutine that would pop 2 goroutines and would wait until nbHeavyRequestsInflight == 2 so that the rest of the test can continue